### ISSUE TITLE
Inserting date

### ISSUE THREAD

**Issue by FredericLatour** [sidorares/node-mysql2#712]
There must be something on my side but I am somewhat stuck.

I'm inserting data into a table using the following approach:

```javascript
const p = await getPool(name); // caching the pool because I'm accessing multiple dbs
const insertFbl = 'INSERT INTO fbl SET ?'
const result = p.query(insertFbl, some_object)
```
`some_object` has some date fields.

When I run the query I get an error like this (for the sake of testing I did init my date field with new Date()):

Incorrect datetime value: '2018-01-21T00:37:07.837Z' for column 'feedback_date' at row 1
    at PromisePool.query (/server/node_modules/mysql2/promise.js:344:20)
    ...

Any idea?

**Comment by FredericLatour** [comment]
Ok, let's try to rephrase this another way.

Can anyone confirm that the following query should work properly providing that `fbl` table has a `some_date` DATETIME field?

```javascript
const myData = {some_date: new Date()}
const result = pool.query('INSERT INTO fbl SET ?', myData)
```
Thanks

**Comment by dougwilson** [comment]
Yea, that works for me. The SQL that it creates (for me right now in my time zone) is:
```sql
INSERT INTO fb1 SET `some_date` = '2018-01-26 20:14:26.012'
```

**Comment by dougwilson** [comment]
In your above post, I see the error is saying you're date looks like `'2018-01-21T00:37:07.837Z'` which is not a string that MySQL server recognizes as a date. I think it's the `Z` character at the end that is throwing it off:
```
mysql> SELECT CONVERT('2018-01-21T00:37:07.837Z', DATETIME);
+-----------------------------------------------+
| CONVERT('2018-01-21T00:37:07.837Z', DATETIME) |
+-----------------------------------------------+
| 2018-01-21 00:37:07.837000                    |
+-----------------------------------------------+
1 row in set, 1 warning (0.00 sec)

Warning (Code 1292): Truncated incorrect datetime value: '2018-01-21T00:37:07.837Z'
mysql> SELECT CONVERT('2018-01-21T00:37:07.837', DATETIME);
+----------------------------------------------+
| CONVERT('2018-01-21T00:37:07.837', DATETIME) |
+----------------------------------------------+
| 2018-01-21 00:37:07.837000                   |
+----------------------------------------------+
1 row in set (0.00 sec)
```

**Comment by FredericLatour** [comment]
Thanks for getting back to me.
The weird thing is that I simply used a javascript date object. The conversion should have taken place automatically. Instead of that I got the error displayed in my first post.

**Comment by dougwilson** [comment]
Yea, that is weird. I used the code and it converted just fine. Is it possible you can provide the full code you're using, including constructing your object in the code so we can both have the exact same object to poke at?

**Comment by FredericLatour** [comment]
The code involves serveral modules and stuff. I was hoping that maybe someone had faced this weird behaviour.
Let me first do my homework and come up with a simplified sample based on the same table and the same object.
That will be easier for you do dig in and maybe it will help me find what's going wrong.

**Comment by dougwilson** [comment]
Awesome 👍 I'll I'm looking for is the simplest example that has the error you're getting above. The example in https://github.com/sidorares/node-mysql2/issues/712#issuecomment-360946986 didn't produce any error when I ran it after creating a table with a single column so it would run.

**Comment by midnightcodr** [comment]
I am wondering if it's safe (in terms of SQL injection vulnerability) to write code like 
```
p.query(insertFbl, some_object)
```
as `query` does not do prepared statement like `execute` do.

**Comment by sidorares** [comment]
@midnightcodr interpolating sql client side does require more attention but thus far there is no known vulnerabilities with correct usage of `SqlString.escape()`/`SqlString.format()` ( both mysql and mysql2 use https://www.npmjs.com/package/sqlstring for client side PS-like query building )

**Comment by midnightcodr** [comment]
@sidorares Just to confirm, does that mean `query` use `SqlString.escape()` and `SqlString.format()` in the back when invoking it like `query(insertFbl, some_object)`

**Comment by sidorares** [comment]
yes, also see links to sqlstring source where is handles `"?", [{key: value}]` interpolation https://github.com/sidorares/node-mysql2/issues/553#issuecomment-437221838

**Comment by midnightcodr** [comment]
Thanks @sidorares . Learned something new (and cool) about doing prepared stmt in cli.
